docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,371 @@
+# ADR 64: ABCI 2.0 Integration (Phase II)
+
+## Changelog
+
+* 2023-01-17: Initial Draft (@alexanderbez)
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR outlines the continuation of the efforts to implement ABCI++ in the Cosmos
+SDK outlined in [ADR 060: ABCI 1.0 (Phase I)](adr-060-abci-1.0.md).
+
+Specifically, this ADR outlines the design and implementation of ABCI 2.0, which
+includes `ExtendVote`, `VerifyVoteExtension` and `FinalizeBlock`.
+
+## Context
+
+ABCI 2.0 continues the promised updates from ABCI++, specifically three additional
+ABCI methods that the application can implement in order to gain further control,
+insight and customization of the consensus process, unlocking many novel use-cases
+that previously not possible. We describe these three new methods below:
+
+### `ExtendVote`
+
+This method allows each validator process to extend the pre-commit phase of the
+CometBFT consensus process. Specifically, it allows the application to perform
+custom business logic that extends the pre-commit vote and supply additional data
+as part of the vote, although they are signed separately by the same key.
+
+The data, called vote extension, will be broadcast and received together with the
+vote it is extending, and will be made available to the application in the next
+height. Specifically, the proposer of the next block will receive the vote extensions
+in `RequestPrepareProposal.local_last_commit.votes`.
+
+If the application does not have vote extension information to provide, it
+returns a 0-length byte array as its vote extension.
+
+**NOTE**: 
+
+* Although each validator process submits its own vote extension, ONLY the *proposer*
+  of the *next* block will receive all the vote extensions included as part of the
+  pre-commit phase of the previous block. This means only the proposer will
+  implicitly have access to all the vote extensions, via `RequestPrepareProposal`,
+  and that not all vote extensions may be included, since a validator does not
```

Comment:
naive question: since nodes have the data but the entire valset, which would be included in the next proposal is there a way to save what the validator has seen already?